### PR TITLE
Initial commit of auido and i2s driver

### PR DESCRIPTION
**Pull request by btheosam**

### CODE REVIEW

File: os/drivers/audio/Kconfig [hunk 1]
```diff
@@ -0,0 +1,126 @@
+#
+# For a description of the syntax of this configuration file,
+# see the file kconfig-language.txt in the NuttX tools repository.
```
Comment (sunghan-chang) [inline os/drivers/audio/Kconfig]: NuttX, so let's change this line.
# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt

File: os/drivers/audio/Make.defs [hunk 2]
```diff
@@ -0,0 +1,77 @@
+###########################################################################
+#
+# Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline os/drivers/audio/Make.defs]: 2017

File: os/drivers/audio/alc5658.c [hunk 3]
```diff
@@ -0,0 +1,1736 @@
...
+		return 0;
+	}
+
+	ret = i2c_read(dev, alc5658_i2c_config, reg, 2);
+	if (ret < 0) {
+		auderr("Error, cannot read reg %x\n", regaddr);
+		return 0;
+	}
+
+	regval = ((uint16_t) reg[0] << 8) | (uint16_t) reg[1];
```
Comment (sunghan-chang) [inline os/drivers/audio/alc5658.c]: remove space after type casting

File: os/include/tinyara/audio/pcm.h [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <tinyara/irq.h>
+
+#ifdef CONFIG_AUDIO_FORMAT_PCM
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/pcm.h]: To use CONFIG, you should include <tinyara/config.h>

File: os/include/tinyara/audio/i2s.h [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+ *   only enqueues the transfer and returns immediately.  Success here only
+ *   means that the transfer was enqueued correctly.
+ *
+ *   When the transfer is complete, a 'result' value will be provided as
+ *   an argument to the callback function that will indicate if the transfer
+ *   failed.
+ *
+ ****************************************************************************/
+
+#define I2S_SEND(d,b,c,a,t) ((d)->ops->i2s_send(d,b,c,a,t))
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/i2s.h]: add space after ,

File: os/include/tinyara/audio/audio_null.h [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <tinyara/irq.h>
+
+#ifdef CONFIG_AUDIO_NULL
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/audio_null.h]: include <tinyara/config.h>

File: os/include/tinyara/audio/alc5658.h [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+ * Included Files
+ ****************************************************************************/
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <tinyara/irq.h>
+#include <tinyara/i2c.h>
+
+#ifdef CONFIG_AUDIO_ALC5658
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/alc5658.h]: !

File: os/drivers/audio/Kconfig [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+#
+config AUDIO_I2SCHAR
+	bool "I2S character driver (for testing only)"
+	default n
+	depends on I2S && AUDIO
```
Comment (sunghan-chang) [inline os/drivers/audio/Kconfig]: Where is "I2S" and "AUDIO"?

**Comment by JunhwanPark** [comment]
Hi @btheosam ,
Do you have the ALC5658 audio codec? I have confirmed that NuttX does not have the ALC5658. Is this your new implementation?

**Pull request closed without merging**